src/c2goto/library/builtin_libs.c

```diff
@@ -6,3 +6,268 @@ int __ESBMC_sync_fetch_and_add(int *ptr, int value)
+#endif
+#endif
+#include <cheri_compressed_cap.h>
+
+#ifdef __ESBMC_CHERI_MORELLO__
+#undef CC_IS_MORELLO
+#endif
+
+#include <cheri/cheric.h>
+// #include <assert.h>
+
+uint64_t __esbmc_clzll(uint64_t v)
+{
+  for(int i = 0; i < 64; i++, v <<= 1)
+    if(v & (uint64_t)1 << 63)
+      return i;
+  __ESBMC_assert(0, "clzll is not defined for zero arguments");
+  return nondet_uint64();
+}
+
+union __esbmc_cheri_cap128
+{
+  void *__capability cap;
+  // struct { __UINT64_TYPE__ pesbt, cursor; };
+  struct
+  {
+    __UINT64_TYPE__ cursor, pesbt;
+  };
+};
+
+__SIZE_TYPE__ __esbmc_cheri_base_get(void *__capability cap)
+{
+  union __esbmc_cheri_cap128 u = {cap};
+  cc128_cap_t comp;
+  cc128_decompress_mem(u.pesbt, u.cursor, true /* tag */, &comp);
+  // __ESBMC_assert(u.comp.cr_bounds_valid, "__esbmc_cheri_base_get on capability with invalid bounds");
+  return comp.cr_base;
+}
+
+__SIZE_TYPE__ __esbmc_cheri_perms_get(void *__capability cap)
+{
+  union __esbmc_cheri_cap128 u = {cap};
+  cc128_cap_t comp;
+  cc128_decompress_mem(u.pesbt, u.cursor, true /* tag */, &comp);
+  // __ESBMC_assert(u.comp.cr_bounds_valid, "__esbmc_cheri_type_get on capability with invalid bounds");
+  __SIZE_TYPE__ hwperms = cc128_get_perms(&comp);
+  __SIZE_TYPE__ swperms = cc128_get_uperms(&comp);
+  return hwperms | swperms << CC128_UPERMS_SHFT;
+}
+
+__UINT16_TYPE__ __esbmc_cheri_flags_get(void *__capability cap)
+{
+  union __esbmc_cheri_cap128 u = {cap};
+  cc128_cap_t comp;
+  cc128_decompress_mem(u.pesbt, u.cursor, true /* tag */, &comp);
+  return cc128_get_flags(&comp);
+}
+
+__UINT32_TYPE__ __esbmc_cheri_type_get(void *__capability cap)
+{
+  union __esbmc_cheri_cap128 u = {cap};
+  cc128_cap_t comp;
+  cc128_decompress_mem(u.pesbt, u.cursor, true /* tag */, &comp);
+  // __ESBMC_assert(comp.cr_bounds_valid, "__esbmc_cheri_type_get on capability with invalid bounds");
+  return cc128_get_otype(&comp);
+}
+
+_Bool __esbmc_cheri_sealed_get(void *__capability cap)
+{
+  union __esbmc_cheri_cap128 u = {cap};
+  cc128_cap_t comp;
+  cc128_decompress_mem(u.pesbt, u.cursor, true /* tag */, &comp);
+  return cc128_is_cap_sealed(&comp);
+}
+
+/* modelled after UCAM-CL-TR-951 semantics of CHERI-MIPS instruction CSetBounds */
+void *__capability
+__esbmc_cheri_bounds_set(void *__capability cap, __SIZE_TYPE__ sz)
+{
+#if 1
+  union __esbmc_cheri_cap128 u = {cap};
+  cc128_cap_t comp;
+  cc128_decompress_mem(u.pesbt, u.cursor, true /* tag */, &comp);
+  __PTRADDR_TYPE__ cursor = comp._cr_cursor;
+  __PTRADDR_TYPE__ base = comp.cr_base;
+  __PTRADDR_TYPE__ top = comp._cr_top;
+#else
+  __PTRADDR_TYPE__ cursor = (__PTRADDR_TYPE__)cap;
+  __PTRADDR_TYPE__ base = cheri_getbase(cap);
+  __PTRADDR_TYPE__ top = cheri_gettop(cap);
+#endif
+  __ESBMC_assert(cheri_gettag(cap), "tag-violation c2exception");
+  // __ESBMC_assert(!cc128_is_cap_sealed(&comp) /*cheri_getsealed(cap)*/, "seal-violation c2exception");
```

Comment:
Do you know if we need to keep this comment code here?

Reply:
Not needed, removed.